doc/tutorials/highlevel.md

```diff
@@ -0,0 +1,322 @@
+
+/* file header ... */
+/* helper macros ... */
+
+/**
+* Tag name for 'mydouble'
+*
+*/
+#define ELEKTRA_TAG_MYDOUBLE Mydouble
+
+/**
+* Tag name for 'myfloatarray/#'
+*
+* Required arguments:
+*
+* - kdb_long_long_t index1: Replaces occurence no. 1 of # in the keyname.
+*
+*
+*/
+#define ELEKTRA_TAG_MYFLOATARRAY Myfloatarray
+
+/* local macros ... */
+
+static inline kdb_double_t ELEKTRA_GET (Mydouble) (Elektra * elektra) { /* ... */ }
+static inline void ELEKTRA_SET (Mydouble) (Elektra * elektra, kdb_double_t value, ElektraError ** error) { /* ... */ }
+
+static inline kdb_float_t ELEKTRA_GET (Myfloatarray) (Elektra * elektra, kdb_long_long_t index1) {  /* ... */ }
+static inline void ELEKTRA_SET (Myfloatarray) (Elektra * elektra, kdb_float_t value, kdb_long_long_t index1, ElektraError ** error) { /* ... */ }
+
+/* local macros ... */
+
+int loadConfiguration (Elektra ** elektra, ElektraError ** error);
+void printHelpMessage (Elektra * elektra, const char * usage, const char * prefix);
+void specloadCheck (int argc, const char ** argv);
+
+/* elektra* macros ... */
+```
+
+At the top of the file (after some helper macros) we can see all the `ELEKTRA_TAG_*` macros. We use these to allow for some flexibility in
+how we generate the names of the `static inline` functions further down. You should always refer to your config values via these macros.
+Additionally, the comments for these macros contain the documentation on what arguments are needed for accessing the tag in question. For
+example to access the elements of the array `myfloatarray/#`, we obviously need to provide an index.
+
+Next we see some local helper macros only used in this file.
+
+Then the main part of the header: For each config value we generate an `ELEKTRA_GET(*)` and an `ELEKTRA_SET(*)` accessor function. All these
+functions are `static inline`, because they just call other getter/setter functions with partially fixed arguments. In fact many of these
+functions will only be a single line.
+
+Then we undefine the local macros we defined before and declare the three initialization functions `loadConfiguration`, `printHelpMessage`
+and `specloadCheck`.
+
+At the end of the file you will find the `elektra*` convenience macros. These macros can be used to make accessing config values look more
+like normal function calls and avoid the ugly double parentheses in e.g. `ELEKTRA_GET (...) (...)`.
+
+### Obtaining an `Elektra` handle
+
+We start at the bottom of our `conf.h` excerpt. `specloadCheck` is used to initiate specload mode, if needed. This mode makes your application
+provide its specification to Elektra. How this works exactly is not so important (see [specload plugin](/src/plugins/specload/README.md)).
+You only need to know, that `specloadCheck` should be called immediately at the start of your `main` function and that it only returns, when
+your application is not in specload mode.
+
+```c
+int main (int argc, const char ** argv) {
+    specloadCheck (argc, argv);
```

Comment:
It allows a bit more flexibility. Some applications might not want to call `loadConfiguration` immediately after entering `main`.

Reply:
Yes, exactly. This flexibility is what I do not want here: There are many ways how you could call it wrongly or not call at all. This is basically also a configuration option (like the others in the generator we discussed so much),  only that it is hard-coded. The less such options we have, the more streamlined the behavior of applications using Elektra is.

Reply:
> Some applications might not want to call loadConfiguration immediately after entering main.

If they don't want to, they most likely also cannot call specloadCheck directly after main. A concept which requires to do something in main is fundamentally broken.

Reply:
> If they don't want to, they most likely also cannot call specloadCheck directly after main.

`specloadCheck` is very lightweight. It starts with

```
if (argc != 2 || strcmp (argv[1], "--elektra-spec") != 0)
{
    return;
}
```

so it either returns more or less immediately, or it doesn't return at all (as expected in specload mode). If it were integrated in `loadConfiguration` on the other hand, we either don't return (specload mode) or we execute the whole Elektra initialization immediately.

Applications may want to start some other work (independent of the configuration) ASAP (e.g. connect to a server, open a log file, create some worker threads). Doing that before `specloadCheck` isn't wise, because it may not return and resources may leak. On the other hand doing it after `loadConfiguration` may cause unwanted start-up delays.

> A concept which requires to do something in main is fundamentally broken.

Yes, we already discovered that the concept of `specload` is not the best.

Reply:
Yes, I am afraid you are right. There will be situations where `loadConfiguration` is called too late.  E.g. applications might modify the output streams and garbage gets printed (or nothing) instead of the specification. But even if someone needs to calls it earlier, we still can call it within `loadConfiguration`, as it can never harm to be called again later.

So we should:

- call it within `loadConfiguration` which should cover 99% of the cases
- `specloadCheck` should be renamed: it does not only check but also do it, e.g. `exitForSpec`?
- for the remaining 1%, we offer a `exitForSpec` but at the end of the tutorial with clear statements when it might be needed (and that it is usually not needed. The cases you gave usually need configuration: to which server to connect, which log file to open and how many working threads to open. But of course it is possible that some applications do this without any configuration from Elektra)

Reply:
I renamed it to `exitForSpecload`, but I did not integrate it into `loadConfiguration`.

The reason is simple. To call `exitForSpecload` you need to pass `argc` and `argv`. That means you need to pass those to `loadConfiguration` as well. This is no problem, if you don't use `exitForSpecload` and call `loadConfiguration` at the start of your `main` function.

But if for some reason you used `exitForSpecload` the best way we could allow calls to `loadConfiguration` is `loadConfiguration(elektra, error, 0, NULL);`. The two extra arguments are just clutter and don't achieve anything. There are also cases (e.g. threaded applications) where `loadConfiguration` needs to be called multiple times, but `exitForSpecload` must only be called once.

Also the documentation would not become simpler. We would still have to explain, why you have to call `loadConfiguration` at the start of your `main` before printing anything to `stdout` and that any resources acquired before calling `loadConfiguration` can only be cleaned up via `atexit`, because `loadConfiguration` may call `exit`.

All of that leads me to the conclusion that it is much better to keep the functions separate.

Reply:
Yes, of course: it would not be a good solution to add argc, argv to loadConfiguration. I was thinking about some solution where this is not needed, like with support from the global plugin gopts (loadConfiguration could add the spec to "spec" and gopts sends whatever is in "spec/"?)?

The problem is that exitForSpecload is a specific hack which is only useful for the specload approach. So when an application developer thinks, he/she does not need the specload feature, he/she will not add the exitForSpecload. And this is a pity in the case that the specification is actually built-in. If we somehow can handle the specload case without forcing the user to add some additional code, the API would be more simple and all applications would be specload-aware.

Reply:
`exitForSpecload` is only generated, if the specification is built-in. So if the function exists, it must be called for correct behaviour. If it doesn't exist, its obvious that we don't call it.

> like with support from the global plugin gopts

Technically this would be possible, although a bit complicated. However, I don't like the solution. First of all it is again far to much magic for my taste. Especially in a language like C, I want to know what exactly all of the code does. Otherwise it is far to easy to create serious flaws.

Right now the fact that `exitForSpecload` expects `argc` and `argv` makes it pretty obvious that it has to be called in `main`. It might still be called to late, but that cannot be prevented (except with good documentation).

Without the explicit arguments, it would be even easier to miss the very strict requirements for the function handling `specload` (whichever function that may be).

Reply:
One option would be to have a static field that is set by the `exitForSpecload` and checked by `loadConfiguration`. If it the field is not set when `loadConfiguration` is called, we immediately return an error.

Reply:
> exitForSpecload is only generated, if the specification is built-in. So if the function exists, it must be called for correct behaviour. If it doesn't exist, its obvious that we don't call it.

This makes even more a cliff between specload-aware applications and others that are not. I think you should always generate the function, even if it does nothing. Then making an application specload-aware is only a build-option and not a source-code change (if exitForSpecload was added correctly already).

> One option would be to have a static field that is set by

I do not like the idea of enforcing  something that is actually a problem in the API, it only creates even more problems (e.g. plugins using loadConfiguration would fail if used from an application that does not call exitForSpecload).

Furthermore, to do it properly we would need pthreads for a mutex (otherwise we would produce potential race conditions).

> First of all it is again far to much magic for my taste.

I understand the concerns that it is kind of "magical" what gopts does in that case of exitForSpecload but gopts is already past the point of being easily understandable. But in any case it is understandable for people that are interested, adding one more if for exitForSpecload would not make a big difference.

But of course at some point the line needs to be drawn, e.g., the next step would be shouldn't we also add printHelpMessage within gopts by default to get uniform behavior between applications?

Furthermore, a big point against exitForSpecload integrated in gopts is that plugins in Elektra actually should not call exit.

Maybe you can write down what we discussed here in an decision? (So that there is some rationale for exitForSpecload, and maybe you find one more point for/against it in the process.)

Reply:
> Then making an application specload-aware is only a build-option and not a source-code change (if exitForSpecload was added correctly already).

If someone wants an application that supports both embedded and external spec, they can always add an `#ifdef` with a macro injected by the build system.

I think it is much more confusing to have an `exitForSpecload` function, when `specload` is not used.

> already past the point of being easily understandable

`gopts` is very easy to understand. It uses system specific APIs to retrieve `argc`, `argv` and `envp` and then calls `elektraGetOpts`. To understand the system specific APIs you obviously need some understanding of the system. But even then apart from win32 the implementations are pretty simple.

>  the next step would be shouldn't we also add printHelpMessage within gopts by default to get uniform behavior between applications?

What do you mean by that?

> Furthermore, a big point against exitForSpecload integrated in gopts is that plugins in Elektra actually should not call exit.

When did we start talking about putting `exitForSpecload` into `gopts`? That makes no sense whatsoever. `gopts` is completely independent of any code-generated functions. The name `exitForSpecload` can also be changed by the user to match possible naming conventions.

Reply:
>  If someone wants an application that supports both embedded and external spec, they can always add an #ifdef with a macro injected by the build system.

I do not like the idea that one needs to support some kind of spec by changing code. The code should always be the same, only the options to the code generator should choose what is supported.

> I think it is much more confusing to have an exitForSpecload function, when specload is not used.

As application developer you do not know if specload will be used by the distributions. The whole point of Elektra is to move such decisions to distributions/end users. Think of specload as yet-another configuration-file format.

> When did we start talking about putting exitForSpecload into gopts?

Integrating does not necessarily mean to put it there. Of course it would be decoupled so that gopts would simply output whatsoever is in spec.

> The name exitForSpecload can also be changed by the user to match possible naming conventions.

I am also not so thrilled about that feature as it does not really offer what someone would expect. The non-generated high-level API would still not match. But this feature does not look harmful.

Reply:
> The code should always be the same, only the options to the code generator should choose what is supported.

This does work here, but is not a good approach in general. It is a **code** generator. One has to expect that changing its arguments might require changes in other code.

> As application developer you do not know if specload will be used by the distributions.

The decision here is not whether to use `specload` or not, but whether to embed the specification. `specload` can be used without embedding the specification and this is also mentioned in the documentation.

> I am also not so thrilled about that feature as it does not really offer what someone would expect. The non-generated high-level API would still not match. But this feature does not look harmful.

The difference here is the non-generated API actually is an API and delivered as a shared library. It is completely external from the point of an application developer. The code-generated files are delivered as source-code and are part of the application itself.

Also, if we were to remove the renaming feature, we would need to add some `elektra` prefix to the names to avoid name conflicts that way.

---

To some this discussion up:

- I will change the templates, so that `exitForSpecload` is always generated. Although I still find it weird to generate empty and useless functions.
- I will try to sum up this discussion in a decision.

Reply:
Thank you! :sparkle: